### PR TITLE
Build vagrant from source

### PR DESCRIPTION
**Pull request by aneeshusa**
###### Motivation for this change

I got tired enough of all the hacks in the Vagrant derivation to take an excursion into Ruby-land packaging. The end result isn't perfect, but IMO it's a great deal better.
See individual commit messages for details.

Other highlights:
- New `url` source type for Ruby gems
- Probably a mass-rebuild of Ruby-dependent packages (not sure)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc maintainers @lovek323 @globin @jgeerds @kamilchm 
cc @zimbatm and @cstrahan as Ruby maintainers
cc @samueldr: I tried to preserve the completions from #30482, but didn't know how to test this

### CODE REVIEW

**Review by zimbatm — COMMENTED** [review]
Hey, first of all this is awesome. I am glad to see the omnibus go away.

Would you mind submitting the ruby changes separately so that they can be reviewed more easily? Once this is merged the vagrant package would go in next.

**Comment by aneeshusa** [comment]
Rebased for Vagrant 2.0.1, which has native Virtualbox 5.2 support.
cc @zimbatm I've opened #33014 to split out the first three commits here with general Ruby changes for you to review more easily.
I've kept the 4th commit here, which is also a Ruby infra change that adds the new `url` `source` type. I basically just changed enough bits in that commit to make the Vagrant derivation (in the last commit) work, so the two are fairly coupled and I think it will be easier to review with that context available, as well as easier to make any improvements you suggest.

File: pkgs/development/tools/vagrant/default.nix [hunk 1]
```diff
@@ -1,147 +1,64 @@
...
+  patches = [
+    (fetchurl {
+      name = "vagrant-backport-virtualbox-5.2-support.diff";
+      url = "https://github.com/hashicorp/vagrant/commit/7d73af5637de41f1e53b8f1ef2ea9baf76842dfb.diff";
+      sha256 = "0dv80fmfrib8gj8priiqa9b2d5mrkgk7c3jbs97xxjwh5xga9sil";
+    })
+  ];
+
+  postInstall = let
+    # Why is this 2.3.0 and not ${ruby.version}?
```
[the remark this answers is not in the document]
Comment (aneeshusa) [reply pkgs/development/tools/vagrant/default.nix]: I changed this to `${ruby.version.libDir}`, which appears to be what I want here.

**Comment by aneeshusa** [comment]
Also cc @ryanartecona @zzamboni @nixy: I'd love if you could test this branch out on Darwin if you get a chance, as I can't test that myself!

**Comment by zzamboni** [comment]
@aneeshusa thanks for this - unfortunately I have not been able to test due to https://github.com/NixOS/nix/issues/1583, which causes my Mac to freeze whenever I try to compile anything heavy-duty. I will test when I'm able to upgrade to 10.13.2 (my Mac is centrally managed at work, so I cannot upgrade yet) and report back.

**Comment by ryanartecona** [comment]
Just tested this locally, and it works great for me! I got the result `/nix/store/z0snfc044b5ha673s5pkqic7nwpdqc51-vagrant`. I've tested it with `vagrant global-status` and bringing an existing vagrant VM I had up and down.

The "You appear to be running Vagrant outside of the official installers" warning is pretty annoying, but agreed on not trying too hard to hack around it. We could get it to stop printing by setting `VAGRANT_INSTALLER_ENV=1` in a wrapper script, but that seems to have other implications inside vagrant code which will make it look inside `$VAGRANT_INSTALLER_EMBEDDED_DIR` for some stuff, and I don't understand those in depth. If anything, I imagine we could maintain a small patch to the vagrant executable to just comment out the line that prints that warning, so Nix users don't see an unnecessary warning all the time.

**Comment by zzamboni** [comment]
@aneeshusa I upgraded my Mac to 10.3.2 this morning so I now had a chance to test this branch - it seems to work perfectly well. As @ryanartecona mentioned, I get a warning about not running Vagrant from an official installer. I have for now suppressed it by setting `VAGRANT_INSTALLER_ENV=1` in my environment - so far no side effects are apparent, but I will keep an eye on it and report back if anything breaks.

Thanks for your work on this!

**Comment by aneeshusa** [comment]
Thanks for testing @ryanartecona @zzamboni!
I've rebased this on the a freshly-rebased #33014, and added a patch to not show the unofficial installation warning. See the commit message for details on why that is better than setting the `VAGRANT_INSTALLER_ENV` variable.

**Comment by zimbatm** [comment]
thanks @aneeshusa. I merged the main PR, let's wait 1 day to see how this goes. Ping me if I forget this PR.

**Comment by ryanartecona** [comment]
> Ping me if I forget this PR.

ping @zimbatm :)

**Comment by zimbatm** [comment]
thanks :)

**Pull request merged**